Fixes on rule link_baseline_coexpression

pmb59 opened the pull request.

pmb59 committed:
WIP on rule link_baseline_coexpression

pinin4fjords committed:
Fix linking rule, reduce file system globbing

pmb59 committed:
Includes checkpoint and def input_for_link

pinin4fjords committed:
Remove unecessary error

pinin4fjords commented:
@pmb59 I've applied a putative fix for the symlinking once I understood the logic better. It will need testing but you should see what I mean. Basically we can just use what we already know from get_metrics() to make that rule work. I also made it so get_metrics() is only run once so we don't keep globbing the file system (naughty @pcm32).

pinin4fjords commented on Snakefile:
```diff
@@ -82,6 +82,7 @@ def get_metrics():
                 metric_grabbed.append(metric)
```
We should add a check here that at least one metric is available

pmb59 replied:
See  d84a8a2

pinin4fjords committed:
Remove excess mkdir -p

pinin4fjords commented on Snakefile:
```diff
@@ -146,9 +150,10 @@ def get_outputs():
+        outputs.extend(expand(f"{config['accession']}"+"-heatmap-{metric}.pdf", metric=metrics ))
     if 'baseline-coexpression' in config['tool'] or 'all-baseline' in config['tool'] and skip(config['accession'],'baseline-coexpression'):   
-        outputs.extend(expand(f"{config['accession']}"+"-{metric}-coexpressions.tsv.gz", metric=get_metrics() )) 
+        outputs.extend(expand(f"{config['accession']}"+"-{metric}-coexpressions.tsv.gz", metric=metrics )) 
```
This needs to be modified to check the number of columns in the expression files (TPM and FPKM should have the same number of columns), and only request the coexpression files if the column number is greater than 3. 

This is a lot simpler to do than trying to account for dynamic failures in the baseline_coexpression rule.

pinin4fjords replied:
FYI @pcm32

pmb59 replied:
Do you mean simpler than using a checkpoint in `baseline_coexpression`?

pinin4fjords replied:
Checkpoint is overly complex for situations where we know in advance the output files. It's very easy here to read the first line of the TSVs  for FPKM and TPM and decide based on that whether or not we can execute the coexpression steps. It would also be wasteful of resources to submit cluster jobs for this when we know in advance they will fail.

pmb59 committed:
Revert "Includes checkpoint and def input_for_link"

This reverts commit 66faf1921101d398ecbb546bc1f8863f93e83d37.

pmb59 committed:
Merge branch 'feature/link-coexp' of https://github.com/ebi-gene-expression-group/bulk-recalculations into feature/link-coexp

pmb59 committed:
Includes a control mechanism to require outputs for 'rule baseline_coexpression' and 'rule link_baseline_coexpression' only when the metric (fpkms or tpms) is present and when the dataset has more than 2 data columns.

pmb59 committed:
Corrects variable naming in `def check_config_required`

pmb59 committed:
Detects if any metric is available for the experiment, otherwise exits.

pcm32 commented on Snakefile:
```diff
@@ -329,22 +348,14 @@ rule link_baseline_coexpression:
-            echo "Error: neither TPM nor FPKM coexpressions.tsv.gz file found"
-            exit 1
-        fi
+	    ln -s {input} {output}
```
```suggestion
	    ls -l {input)
	    ln -s {input} {output}
	    ls -l {input}
	    ls -l {output}
```

pcm32 commented on Snakefile:
```diff
@@ -146,9 +162,17 @@ def get_outputs():
-        outputs.extend(expand(f"{config['accession']}"+"-{metric}-coexpressions.tsv.gz", metric=get_metrics() )) 
+        metric_link_coexp=False
+        for m in metrics:
+            if os.path.getsize(f"{config['accession']}"+"-"+m+".tsv") > 0 and get_number_columns(f"{config['accession']}"+"-"+m+".tsv")>4:
```
```suggestion
            expression_file=f"{config['accession']}-{m}.tsv"
            print(f"Checking file size for {expression_file} and number of columns")
            if os.path.getsize(expression_file) > 0 and get_number_columns(expression_file)>4:
```

pcm32 commented on Snakefile:
```diff
@@ -146,9 +162,17 @@ def get_outputs():
+        for m in metrics:
+            if os.path.getsize(f"{config['accession']}"+"-"+m+".tsv") > 0 and get_number_columns(f"{config['accession']}"+"-"+m+".tsv")>4:
+                metric_link_coexp=True
+                outputs.extend(expand(f"{config['accession']}"+"-"+m+"-coexpressions.tsv.gz"))
```
```suggestion
                outputs.extend(expand(f"{config['accession']}-{m}-coexpressions.tsv.gz"))
```

pmb59 committed:
Few updates on `def get_outputs` and 'link_baseline_coexpression'

pmb59 committed:
Defines function `read_delete_previous_output()`

pmb59 committed:
* sets `DELETE_PREV_OUTPUT` to False
* Sets `--forceall` in Snakefile-sorting-hat
	- Force the execution of the selected first rule and all rules it is dependent on regardless of already created output.

pmb59 committed:
* Moves metrics as a global variable inside `def get_outputs()`
* `metrics` is only defined when required (baseline)
* Reinstates lambda function for rule `link_baseline_coexpression`

pcm32 commented on Snakefile:
```diff
@@ -329,22 +362,17 @@ rule link_baseline_coexpression:
-            echo "Error: neither TPM nor FPKM coexpressions.tsv.gz file found"
-            exit 1
-        fi
+        ls -l {input}
+        ln -s {input} {output}
+        ls -l {input}
+        ls -l {output}
```
```suggestion
        ln -s {input} {output}
```

we can remove these checks and go back to this now.

pcm32 commented on Snakefile-sorting-hat:
```diff
@@ -187,7 +187,7 @@ rule produce_recalculations_call:
 
 
-        snakemake --latency-wait 100 {config['sm_options']} --config \
+        snakemake --forceall --latency-wait 150 {config['sm_options']} --config \
```
`--forceall` needs to be optional, so that if a variable is set, then it runs with it. We need to be able to turn this on and off. It will only be used like this on the first run of the release, then subsequent runs of the same release shouldn't have to re-run everything.

pcm32 commented on Snakefile:
```diff
@@ -146,20 +174,30 @@ def get_outputs():
-    
-    if 'delete_previous_output' in config and config['delete_previous_output']==True:
+    print(datetime.datetime.now())
+    if read_delete_previous_output():
         for x in outputs:
             print('Trying to delete existing output: '+ x)
             try:
                 os.remove(x)
             except:
                 print("Output file ", x, " not found in ", os.getcwd())
```
```suggestion
```

If using `--forceall` this can go now.

pmb59 committed:
Update Snakefile (removes ls -l)

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
Update Snakefile (removes if read_delete_previous_output(): )

Co-authored-by: Pablo Moreno <pcm32@users.noreply.github.com>

pmb59 committed:
- removes option delete_previous_output

pmb59 committed:
Merge branch 'feature/link-coexp' of https://github.com/ebi-gene-expression-group/bulk-recalculations into feature/link-coexp

pmb59 committed:
removes get_delete_previous_output()

pmb59 committed:
Removes `delete_previous_output` PARAM in rule produce_recalculations_call

pmb59 committed:
Removes `read_delete_previous_output()` in Snakefile

pmb59 committed:
Merge branch 'develop' into feature/link-coexp

pmb59 committed:
Update run_sorting_hat_test_data.sh

pcm32 commented on run_sorting_hat_test_data.sh:
```diff
@@ -1,12 +1,20 @@
 SORTING_HAT=$( pwd )/Snakefile-sorting-hat
 LOG_HANDLER=$( pwd )/log_handler.py
 SN_CONDA_PREFIX=${SN_CONDA_PREFIX:-$( pwd )/conda_installs}
+if [ $FORCEALL ]
+then
+  PROFILE_LINE="--profile profilename --restart-times "$RESTART_TIMES
+else
+  PROFILE_LINE="--profile profilename --forceall --restart-times "$RESTART_TIMES
+fi
```
I wouldn't mix all these things together here... if forceall should only change a $FORCEALL_LINE or something alike.

pmb59 replied:
See ddce3e9

pcm32 commented on Snakefile-sorting-hat:
```diff
@@ -190,12 +183,11 @@ rule produce_recalculations_call:
 
 
-        snakemake --forceall --latency-wait 150 {config['sm_options']} --config \
+        snakemake --latency-wait 150 {config['sm_options']} --config \
```
How are we passing the forceall behaviour to the Snakefile analysis part? When the sorting hat receives the command to force, then the subworkflow needs to apply it as well, otherwise older results (from a previous release, won't be re-written).

pmb59 replied:
In the current implementation, `--forceall` is passed (if `FORCEALL=$true`) here `{config['sm_options']}`  to the snakemake call in the sorting hat. See ddce3e9

pcm32 commented on run_sorting_hat_test_data.sh:
```diff
@@ -1,12 +1,20 @@
+FORCEALL=$true
+RESTART_TIMES=1
 SKIP_STEPS=$( pwd )/step_skip.yaml
+CHECK_SPECIES=$( pwd )/Bulk_recalculations_check_species_names.yaml
```
A better name for this file would be `atlas-species-name-mapping.yaml`, since it won't be used only on recalculations.

pmb59 replied:
See c44db20

pcm32 reviewed (commented):
Some comments to help improve this for longer term usage and adoption within the overall processes.

pmb59 committed:
Update run_sorting_hat_test_data.sh

pmb59 committed:
Update run_sorting_hat_test_data.sh

Adds $FORCE_ALL and $RESTART_TIMES to {config['sm_options']}.

pmb59 committed:
Fixes IF statement for --forceall.

pmb59 committed:
Prints time taken by the main snakemake call to complete.

pmb59 committed:
Merge branch 'develop' into feature/link-coexp

pmb59 committed:
Sets up env variables for maximum number of cluster jobs that can be run by Snakemake

pcm32 commented on run_sorting_hat_test_data.sh:
```diff
@@ -1,20 +1,20 @@
+NJOBS=${NJOBS:-10}
 GTF=$( pwd )/test-data/gff
-FORCEALL=$true
+FORCEALL=true
```
```suggestion
FORCEALL=${FORCEALL:-false}
```
I would leave the least dangerous case as default.

pcm32 replied:
On second thoughts, this is not part of the PR, it was probably resolved elsewhere.... I'll remove the commit suggestion. Feel free to merge as it is.

pcm32 reviewed (approved):
LGTM

pmb59 committed:
sets default value for FORCEALL
